File: proga/1_semestr/labs/lab3/src/main/java/dev/lysmux/lab3/Main.java
package dev.lysmux.lab3;

import dev.lysmux.lab3.common.Case;
import dev.lysmux.lab3.entity.Glow;
import dev.lysmux.lab3.entity.Substance;
import dev.lysmux.lab3.entity.ability.EmitRays;
import dev.lysmux.lab3.entity.ability.GlowsInDark;
import dev.lysmux.lab3.entity.item.MoonStone;
import dev.lysmux.lab3.entity.people.Person;
import dev.lysmux.lab3.entity.people.ShortyGroup;
import dev.lysmux.lab3.entity.place.Nature;
import dev.lysmux.lab3.entity.place.Table;
import dev.lysmux.lab3.entity.ray.Ray;
import dev.lysmux.lab3.entity.ray.RayKind;
import dev.lysmux.lab3.entity.ray.VisibilityType;
import dev.lysmux.lab3.storage.IStoryStorage;
import dev.lysmux.lab3.storage.StoryStorageImpl;
import dev.lysmux.lab3.storage.exception.StoryExistsException;
import dev.lysmux.lab3.story.SentenceFactory;
import dev.lysmux.lab3.story.Story;
import dev.lysmux.lab3.story.exception.EmptyStoryException;
import dev.lysmux.lab3.util.Direction;
import dev.lysmux.lab3.util.Util;

public class Main {
    public static void main(String[] args) {
        Story story = new Story();

        Person person = new Person();
        ShortyGroup shortyGroup = new ShortyGroup("рассевшимися вокруг");

        MoonStone moonStone = new MoonStone();
        Table table = new Table();
        Glow glow = new Glow();

        Substance glows_substance = new Substance(new GlowsInDark());
        Substance rays_substance = new Substance(new EmitRays(
                new Ray(RayKind.LIGHT, VisibilityType.VISIBLE),
                new Ray(RayKind.ULTRAVIOLET, VisibilityType.HIDDEN),
                new Ray(RayKind.INFRARED, VisibilityType.HIDDEN),
                new Ray(RayKind.COSMIC, VisibilityType.HIDDEN)
        ));

        story.addSentences(
                SentenceFactory.createSentence(
                                Util.capitalize(person.toString()),
                                person.put(
                                        moonStone,
                                        table,
                                        new Direction(
                                                Direction.Preposition.IN_FRONT_OF,
                                                shortyGroup.getDescription()
                                        )
                                )
                        )
                        .and(person.tell())
                        .that(glows_substance.meet(new Nature()))
                        .which(glows_substance.acquireAbility())
                        .after(glows_substance.influenced(new Ray(RayKind.LIGHT))),
                SentenceFactory.createSentence(
                        "Такое",
                        glow.toString(),
                        "называется",
                        glow.getAlternativeName()
                ),
                SentenceFactory.createSentence(
                        "Некоторые",
                        rays_substance.caseDeclension(Case.NOMINATIVE),
                        rays_substance.acquireAbility()
                )
        );

        try {
            System.out.println(story.makeStory());
        } catch (EmptyStoryException e) {
            System.err.println(e.getMessage());
        }

        System.out.println("Сохраняем рассказ...");
        IStoryStorage storage = StoryStorageImpl.getInstance();
        try {
            storage.save("Some рассказ", story);
        } catch (StoryExistsException e) {
            System.err.println(e.getMessage());
        }
    }
}

File: proga/1_semestr/labs/lab3/src/main/kotlin/dev/lysmux/lab3/Main.kt
package dev.lysmux.lab3

import dev.lysmux.lab3.common.Case
import dev.lysmux.lab3.entity.Glow
import dev.lysmux.lab3.entity.Substance
import dev.lysmux.lab3.entity.ability.EmitRays
import dev.lysmux.lab3.entity.ability.GlowsInDark
import dev.lysmux.lab3.entity.item.MoonStone
import dev.lysmux.lab3.entity.people.Person
import dev.lysmux.lab3.entity.people.ShortyGroup
import dev.lysmux.lab3.entity.place.Nature
import dev.lysmux.lab3.entity.place.Table
import dev.lysmux.lab3.entity.ray.Ray
import dev.lysmux.lab3.entity.ray.RayKind
import dev.lysmux.lab3.entity.ray.VisibilityType
import dev.lysmux.lab3.storage.StoryStorageImpl
import dev.lysmux.lab3.storage.exception.StoryExistsException
import dev.lysmux.lab3.story.Story
import dev.lysmux.lab3.story.exception.EmptyStoryException
import dev.lysmux.lab3.util.Direction
import dev.lysmux.lab3.util.strCapitalize


fun main() {
    val person = Person()
    val shortyGroup = ShortyGroup("рассевшимися вокруг")

    val moonStone = MoonStone()
    val table = Table()
    val glow = Glow()

    val glowsSubstance = Substance(GlowsInDark())
    val raysSubstance = Substance(
        EmitRays(
            Ray(RayKind.LIGHT, VisibilityType.VISIBLE),
            Ray(RayKind.ULTRAVIOLET, VisibilityType.HIDDEN),
            Ray(RayKind.INFRARED, VisibilityType.HIDDEN),
            Ray(RayKind.COSMIC, VisibilityType.HIDDEN)
        )
    )

    val story = Story.build {
        sentence {
            content(
                strCapitalize(person.toString()),
                person.put(
                    moonStone,
                    table,
                    Direction(
                        Direction.Preposition.IN_FRONT_OF,
                        shortyGroup.getDescription()
                    )
                )
            ) and content(
                person.tell()
            ) that content(
                glowsSubstance.meet(Nature())
            ) which content(
                glowsSubstance.acquireAbility()
            ) after content(
                glowsSubstance.influenced(Ray(RayKind.LIGHT))
            )
        }
        sentence {
            content(
                "Такое",
                glow.toString(),
                "называется",
                glow.getAlternativeName()
            )
        }
        sentence {
            content(
                "Некоторые",
                raysSubstance.caseDeclension(Case.NOMINATIVE),
                raysSubstance.acquireAbility()
            )
        }
    }

    try {
        println(story.asText())
    } catch (e: EmptyStoryException) {
        System.err.println(e.message)
    }

    println("Сохраняем рассказ...")
    try {
        StoryStorageImpl.save("Some рассказ", story)
    } catch (e: StoryExistsException) {
        System.err.println(e.message)
    }
    println("Рассказ сохранён")
}
